### PR TITLE
feat: Invite codes

### PR DESCRIPTION
**Pull request by Mogge**
## 🍰 Pullrequest
Backend part for using invite codes for user registration

### CODE REVIEW

**Comment by ulfgebhardt** [comment]
As far as I can see our change we did together is missing?! @Mogge

**Comment by Mogge** [comment]
What changes do you mean @ulfgebhardt ?

File: backend/src/schema/resolvers/helpers/generateInviteCode.js [hunk 1]
```diff
@@ -0,0 +1,5 @@
+export default function generateInviteCode() {
+  return Array.from({ length: 6 }, (n = Math.floor(Math.random() * 36)) => {
+    return String.fromCharCode(n > 9 ? n + 55 : n + 48)
```
Comment (ulfgebhardt) [inline backend/src/schema/resolvers/helpers/generateInviteCode.js:3]: Here a comment would be awesome ;). Took me a sec to understand.
Also `55` is incorrect - should be `65`. See https://en.wikipedia.org/wiki/ASCII

```suggestion
    // Shift ASCII Characters to numbers (48-57) or A-Z (65-90)
    return String.fromCharCode(n > 9 ? n + 65 : n + 48)
```

File: backend/src/schema/resolvers/inviteCodes.js [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        session.close()
+      }
+    },
+    isValidInviteCode: async (_parent, args, context, _resolveInfo) => {
+      const { code } = args
+      if (!code) return false
+      const session = context.driver.session()
+      const readTxResultPromise = session.readTransaction(async (txc) => {
+        const result = await txc.run(
+          `MATCH (ic:InviteCode { code: $code })
```
Comment (ulfgebhardt) [inline backend/src/schema/resolvers/inviteCodes.js:43]: The user input should be caps-agnostic - so Something like this should do the trick (Pseudocode, not tested)

```suggestion
          `MATCH (ic:InviteCode { code: UPPER($code) })
```

File: backend/src/schema/resolvers/inviteCodes.spec.js [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+      })
+      const code = inviteCode.get('code')
+      const result = await query({ query: isValidInviteCodeQuery, variables: { code } })
+      expect(result.data.isValidInviteCode).toBeFalsy()
+    })
+
+    it('does not validate an invite code which does not exits', async () => {
+      const result = await query({ query: isValidInviteCodeQuery, variables: { code: 'AAA' } })
+      expect(result.data.isValidInviteCode).toBeFalsy()
+    })
```
Comment (ulfgebhardt) [inline backend/src/schema/resolvers/inviteCodes.spec.js:189]:
```suggestion
    })
    
    it('does validate an invite code in lower caps' , async () => {
    ...
    }
```

**Review by ulfgebhardt — CHANGES_REQUESTED** [review]
Awesome work! Just a few pit falls.
- [ ] Ascii encoding incorrect?
- [ ] Lower caps Codes acceptance?

File: backend/src/schema/resolvers/helpers/generateInviteCode.js (hunk 1 above)
Comment (Mogge) [reply backend/src/schema/resolvers/helpers/generateInviteCode.js:3]: No. It is `return String.fromCharCode(n > 9 ? n + 55 : n + 48)` because 10 + 55 = 65

Comment (Mogge) [reply backend/src/schema/resolvers/helpers/generateInviteCode.js:3]: But I agree in commenting this.

File: backend/src/schema/resolvers/inviteCodes.js (hunk 2 above)
Comment (Mogge) [reply backend/src/schema/resolvers/inviteCodes.js:43]: Great. Thank you!

File: backend/src/schema/resolvers/inviteCodes.spec.js (hunk 3 above)
Comment (Mogge) [reply backend/src/schema/resolvers/inviteCodes.spec.js:189]: Yep. This one is missing.

Comment (Mogge) [reply backend/src/schema/resolvers/inviteCodes.spec.js:198]: See line 166

**Review by Tirokk — APPROVED** [review]
Yeah @Mogge 
This is an important peace of work to reach our GO LIVE! 🚀🚀💫

Great and sustainable work you have done.
I suggested just some easy changes and asked some not as important questions …

Wisdom

**Pull request merged**